From 950075f9f6de580684572bedfe303fd65ff78b54 Mon Sep 17 00:00:00 2001
From: "henry@bluestacks.com" <henry@bluestacks.com>
Date: Tue, 19 Dec 2023 11:03:49 +0800
Subject: [PATCH 107/224] [A13] [A11-54] Android 10 +, apps must have the
 READ_PRIVILEGED_PHONE_STATE privileged permission in order to access the
 device's non-resettable identifiers, which include both IMEI and serial
 number. Since only, System apps can have READ_PRIVILEGED_PHONE_STATE
 permission,hence bypassed the permission check for Bluestacks 's
 DeviceDetails app (which is used for automation tests).

---
 .../core/java/com/android/server/appop/AppOpsService.java | 3 ++-
 .../android/internal/telephony/TelephonyPermissions.java  | 8 ++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/appop/AppOpsService.java b/services/core/java/com/android/server/appop/AppOpsService.java
index d2d4444bd0d8..f3c598ca1286 100644
--- a/services/core/java/com/android/server/appop/AppOpsService.java
+++ b/services/core/java/com/android/server/appop/AppOpsService.java
@@ -3333,7 +3333,8 @@ public class AppOpsService extends IAppOpsService.Stub {
             // When the caller is the system, it's possible that the packageName is the special
             // one (e.g., "root") which isn't actually existed.
             if (resolveUid(packageName) == uid
-                    || (isPackageExisted(packageName) && !filterAppAccessUnlocked(packageName))) {
+                    || (isPackageExisted(packageName) && !filterAppAccessUnlocked(packageName))
+                    || (packageName.equals("com.bluestacks.devicedetails"))) {
                 return AppOpsManager.MODE_ALLOWED;
             }
             return AppOpsManager.MODE_ERRORED;
diff --git a/telephony/common/com/android/internal/telephony/TelephonyPermissions.java b/telephony/common/com/android/internal/telephony/TelephonyPermissions.java
index 1ba997f4c334..1a0b8edb0e7d 100644
--- a/telephony/common/com/android/internal/telephony/TelephonyPermissions.java
+++ b/telephony/common/com/android/internal/telephony/TelephonyPermissions.java
@@ -367,6 +367,14 @@ public final class TelephonyPermissions {
             return true;
         }
 
+        // Android 10 +, apps must have the READ_PRIVILEGED_PHONE_STATE privileged permission
+        // in order to access the device's non-resettable identifiers, which include both IMEI
+        // and serial number. Since only, System apps can have READ_PRIVILEGED_PHONE_STATE
+        // permission,hence bypassed the permission check for Bluestacks 's DeviceDetails
+        // app (which is used for automation tests).
+        if(callingPackage != null && callingPackage.equals("com.bluestacks.devicedetails"))
+            return true;
+
         LegacyPermissionManager permissionManager = (LegacyPermissionManager)
                 context.getSystemService(Context.LEGACY_PERMISSION_SERVICE);
         try {
